File: Collections/ReadOnlyListWrapper.vb
Namespace Common

    Public Class ReadOnlyListWrapper(Of T)
        Inherits BaseReadOnlyList(Of T)

        Public Sub New(ByVal List As IList(Of T))
            Me.List = List
        End Sub

        Public Sub New(ByVal List As IReadOnlyList(Of T))
            Me.ROList = List
        End Sub

        Public Overrides ReadOnly Property Count As Integer
            Get
                If Me.List IsNot Nothing Then
                    Return Me.List.Count
                Else
                    Return Me.ROList.Count
                End If
            End Get
        End Property

        Default Public Overrides ReadOnly Property Item(Index As Integer) As T
            Get
                If Me.List IsNot Nothing Then
                    Return Me.List.Item(Index)
                Else
                    Return Me.ROList.Item(Index)
                End If
            End Get
        End Property

        Public Overrides Function GetEnumerator() As IEnumerator(Of T)
            If Me.List IsNot Nothing Then
                Return Me.List.GetEnumerator()
            Else
                Return Me.ROList.GetEnumerator()
            End If
        End Function

        Private ReadOnly List As IList(Of T)
        Private ReadOnly ROList As IReadOnlyList(Of T)

    End Class

End Namespace

File: Collections/ReadOnlyListWrapper.cs
using System.Collections.Generic;

namespace Ks
{
    namespace Common
    {
        public class ReadOnlyListWrapper<T> : BaseReadOnlyList<T>
        {
            public ReadOnlyListWrapper(IList<T> List)
            {
                this.List = List;
            }

            public ReadOnlyListWrapper(IReadOnlyList<T> List)
            {
                this.ROList = List;
            }

            public override int Count
            {
                get
                {
                    if (this.List != null)
                        return this.List.Count;
                    else
                        return this.ROList.Count;
                }
            }

            public override T this[int Index]
            {
                get
                {
                    if (this.List != null)
                        return this.List[Index];
                    else
                        return this.ROList[Index];
                }
            }

            public override IEnumerator<T> GetEnumerator()
            {
                if (this.List != null)
                    return this.List.GetEnumerator();
                else
                    return this.ROList.GetEnumerator();
            }

            private readonly IList<T> List;
            private readonly IReadOnlyList<T> ROList;
        }
    }
}
